### PR TITLE
Deprecate setup.py and move to pyproject.toml

### LINKED ISSUE #6971: Dependencies: deprecate `setup.py` and move to `pyproject.toml`
`pyproject.toml` is the current specification standard for build system requirements for Python projects, defined in PEP-518: https://peps.python.org/pep-0518/#file-format

`setup.py` is not deprecated, but it's recommended to move to `pyproject.toml` if possible:
- https://packaging.python.org/en/latest/discussions/setup-py-deprecated/
- https://stackoverflow.com/questions/75588250/what-are-the-downsides-to-relying-purely-on-pyproject-toml
- https://packaging.python.org/en/latest/discussions/setup-py-deprecated/#is-pyproject-toml-mandatory

To do:
- [ ] Implement this first: https://github.com/rucio/containers/issues/344 (we want to more accurately replicate the package installation process in tests)
- [ ] Follow https://packaging.python.org/en/latest/guides/modernize-setup-py-project/#modernize-setup-py-project to see what can be moved to `pyproject.toml`

### LINKED ISSUE #7677: Remove `client_syntax` checks and `ignoretool` usage
I think this is a leftover from the past, before we had `ruff`. It does exactly the same thing as the `python_ruff` CI check, but it only runs it for a subset of the files.

In the spirit of simplifying the test architecture (#7667), I think we should remove this.

### PR DESCRIPTION
**Pull request by rdimaio**
fixes https://github.com/rucio/rucio/issues/6971

### CODE REVIEW

File: pyproject.client.toml [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+
+dependencies = [
+        'click',
+        'requests',
+        'urllib3',
+        'dogpile-cache<=1.2.2',
+        'packaging',
+        'tabulate',
+        'jsonschema',
+        'dataclasses',
```
Comment (nsoranzo) [inline pyproject.client.toml]:
```suggestion
```

As in https://github.com/rucio/rucio/pull/7665 .

File: pyproject.client.toml [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+name = "rucio-clients"
+dynamic = ["version"]
+
+# For using SSO login option, install these RPM packages: libxml2-devel xmlsec1-devel xmlsec1-openssl-devel libtool-ltdl-devel
+
+dependencies = [
+        'click',
+        'requests',
+        'urllib3',
+        'dogpile-cache<=1.2.2',
```
Comment (nsoranzo) [inline pyproject.client.toml]:
```suggestion
        'dogpile-cache',
```

As in https://github.com/rucio/rucio/pull/7672 .

File: pyproject.server.toml [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+dynamic = ["version"]
+
+dependencies = [
+  'requests<=2.32.3',
+  'urllib3<=1.26.19',
+  'dogpile-cache<=1.2.2',
+  'packaging<=24.1',
+  'tabulate<=0.9.0',
+  'jsonschema<=4.23.0',
+  'dataclasses',  # no upper limit is set in .in or .txt req files
```
Comment (nsoranzo) [inline pyproject.server.toml]:
```suggestion
```

Part of the standard library since Python 3.7 .

File: pyproject.client.toml (hunk 1 above)
Comment (rdimaio) [reply pyproject.client.toml]: Addressed in latest revision

File: pyproject.client.toml (hunk 2 above)
Comment (rdimaio) [reply pyproject.client.toml]: Addressed in latest revision

File: pyproject.server.toml (hunk 3 above)
Comment (rdimaio) [reply pyproject.server.toml]: Addressed in latest revision

**Comment by maxnoe** [comment]
Let me note that while this is a big improvement, it does not fix #7322, which was originally closed as a duplicate of #6971 

So I re-opened that for now.

**Comment by rdimaio** [comment]
> Let me note that while this is a big improvement, it does not fix #7322, which was originally closed as a duplicate of #6971
> 
> So I re-opened that for now.

I agree - I discussed with Martin that we would initially do this change, then possibly a namespace package restructure (https://github.com/rucio/rucio/pull/7577) and then the editable installs would be possible. This is somewhat discussed https://github.com/rucio/rucio/issues/7330 and in theory by resolving https://github.com/rucio/rucio/issues/7330 we'd resolve #7322 too, but it doesn't hurt to have that one open too.

**Comment by bari12** [comment]
Please rebase

File: pyproject.client.toml [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+        'dogpile-cache',
+        'packaging',
+        'tabulate',
+        'jsonschema',
+        'rich',
+        'typing_extensions'
+]
+requires-python = ">=3.9"
+authors = [
+  {name = "Rucio", email = "rucio-dev@cern.ch"},
```
Comment (bari12) [inline pyproject.client.toml]: Change to `rucio-contact` instead (across all files)

**Comment by rdimaio** [comment]
> Please rebase

Rebased in latest revision to fix merge conflicts, will now address your comment

File: pyproject.client.toml (hunk 4 above)
Comment (rdimaio) [reply pyproject.client.toml]: Addressed in latest commit

**Review by bari12 — DISMISSED** [review]

File: pyproject.server.toml [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+
+[tool.setuptools.package-dir]
+"" = "lib"
+rucio = "lib/rucio"
+bin = "bin"
+etc = "etc"
+tools = "tools"
+requirements = "requirements"
+
+[tool.setuptools.packages.find]
```
Comment (maxnoe) [inline pyproject.server.toml:167]: I think this section must also be added to the other two files.

Comment (rdimaio) [reply pyproject.server.toml:167]: Addressed in latest revision

File: tools/generate_version.py [hunk 6]
```diff
@@ -45,13 +45,11 @@ def run_git_command(cmd):
...
-    'final': %s,
-    'version': '%s',
-    'branch_nick': '%s',
-    'revision_id': '%s',
-    'revno': %s
-}
+IS_FINAL = %s
+VERSION = %s
+BRANCH_NICK = %s
+REVISION_ID = %s
```
Comment (Geogouz) [inline tools/generate_version.py]: Aren't these written as bare strings (e.g., VERSION = 38.0.0)? If so, since this is not valid Python syntax, I guess any fresh release build will fail to import `rucio.vcsversion` (the new `pyproject` files declare a dynamic version `attr = "rucio.vcsversion.VERSION"`).

File: pyproject.server.toml [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+
+[build-system]
+requires = ["setuptools>=78.1.0", "wheel>=0.45.1", "packaging"]
+build-backend = "setuptools.build_meta"
+
+[project]
+name = "rucio"
+dynamic = ["version"]
+
+dependencies = [
```
Comment (Geogouz) [inline pyproject.server.toml:23]: Is having here only upper bounds safe?

Comment (maxnoe) [reply pyproject.server.toml:23]: This is quite a big change and weird to do.

The current requirements has == requirements:

https://github.com/rucio/rucio/blob/master/requirements/requirements.server.in

Comment (rdimaio) [reply pyproject.server.toml:23]: I would say so. From [the packaging docks](https://packaging.python.org/en/latest/discussions/install-requires-vs-requirements/#requirements-files), the requirements files are meant to provide a reproducible environment, while the dependencies specified here should be more minimal and not as 'hard-pinned'. We could be more precise and add lower boundaries here, but how can we exactly test which lower boundaries to set for each dependency? I think it would be added complexity for not much gain, particularly since this is mostly deployed as part of a Docker image where `rucio` is likely the main Python package installed, so it can likely get the latest dependencies without conflicts

File: tools/generate_version.py (hunk 6 above)
Comment (rdimaio) [reply tools/generate_version.py]: Good catch, I ran it now with `python tools/generate_version.py 39.0.0` and indeed it'd generate `VERSION = 39.0.0`, which then is seen as a `SyntaxError` at build time. 

Addressed in commit https://github.com/rucio/rucio/pull/7675/commits/b03c8b478d22c58e5042be2245dea051b97e3657, now running `python tools/generate_version.py 39.0.0` results in the following `vcsversion`:

```
'''
This file is automatically generated; Do not edit it. :)
'''
IS_FINAL = True
VERSION = '39.0.0'
BRANCH_NICK = 'patch-6971-pyproject'
REVISION_ID = '7353b74ddb1d43c8e7fe545c5104c61a624d85be'
REVNO = '14212'
```

then running `./tools/build_sdist_wheel.sh all`, packages are built as expected, and looking at the generated `PKG-INFO` the version is parsed as expected:

```
Metadata-Version: 2.4
Name: rucio
Version: 39.0.0
Summary: Rucio server package
Author-email: Rucio <rucio-contact@cern.ch>
Maintainer-email: Rucio <rucio-contact@cern.ch>
License-Expression: Apache-2.0
...
```

File: pyproject.server.toml (hunk 7 above)
Comment (rdimaio) [reply pyproject.server.toml:23]: > This is quite a big change and weird to do.
> 
> The current requirements has == requirements:
> 
> https://github.com/rucio/rucio/blob/master/requirements/requirements.server.in

That's the requirements file, but the current `setuputil.py` uses upper boundaries in the same way:

https://github.com/rucio/rucio/blob/master/setuputil.py#L63-L113

**Comment by Geogouz** [comment]
Shouldn't this PR make `pip install .` work though? I think I am missing the scope of this PR a bit.

**Comment by rdimaio** [comment]
> Shouldn't this PR make `pip install .` work though? I think I am missing the scope of this PR a bit.

The idea is that the different packages should be installed separately https://packaging.python.org/en/latest/guides/packaging-namespace-packages/, and then you can point to those packages as dependencies in the top-level pyproject, like this https://github.com/DIRACGrid/diracx/blob/main/pyproject.toml

However, until the namespace restructure mentioned in https://github.com/rucio/rucio/issues/7330 is done, this cannot be done, so I think we need to replicate the dependencies from the server pyproject into the main pyproject for now as a temporary workaround

**Comment by rdimaio** [comment]
> Shouldn't this PR make `pip install .` work though? I think I am missing the scope of this PR a bit.

Addressed in commit https://github.com/rucio/rucio/pull/7675/commits/42df7ba0465e87e2789b1eab7a13f1f014a03e57

**Review by Geogouz — APPROVED** [review]
Can't find something obvious. I am giving a green light by my side.

**Review by bari12 — APPROVED** [review]
